### PR TITLE
Use `Iterators.Reverse` in `foldr`

### PR DESCRIPTION
**Pull request by simonbyrne**
(as suggested by @TotalVerb)

### CODE REVIEW

**Comment by garrison** [comment]
Why not use `Iterators.reverse` instead?  This will then unwrap any `Reverse` object that is passed.  I believe `Iterators.reverse` is meant to be the API for calling (for instance, it is mentioned in NEWS.md).

**Comment by simonbyrne** [comment]
Good call, I should read the docs more...

**Review by TotalVerb — APPROVED** [review]
lgtm, nice simplification. maybe nanosoldier?

**Comment by stevengj** [comment]
Any chance of rebasing this and running some benchmarks to make sure that there are no performance regressions?

**Comment by stevengj** [comment]
I tried this out manually on my machine, benchmarking on `foo(x) = foldr(-, x)` for `x = rand(1000)`, and it seems to be essentially the same speed as `foldr` on `master`.   However, it is reporting `(1 allocation: 16 bytes)` whereas before it was `(0 allocations: 0 bytes)` … not sure if we care, but it might be nice to avoid this?

I also tried an alternate implementation that avoids the `Switch` type, and it wasn't any faster (nor did it avoid the allocation), so `Switch` doesn't seem to be inhibiting any optimizations.

File: base/reduce.jl [hunk 1]
```diff
@@ -103,30 +103,21 @@ julia> foldl(=>, 1:4)
...
-        end
-        return v
-    end
+# switches the order of arguments of a binary operator
+struct Switch{F}
+    op::F
 end
+(s::Switch)(x,y) = s.op(y,x)
+reduce_empty(s::Switch, T) = reduce_empty(s.op, T)
+reduce_first(s::Switch, x) = reduce_first(s.op, x)
```
Comment (stevengj) [inline base/reduce.jl]: Though this is elegant, it doesn't cover the specialized cases of `mapreduce_empty_iter` that we already have, not to mention future specializations of `mapreduce_first`. 

You can try to override `mapreduce_empty_iter` and `mapreduce_first`, but that seems to be a minefield of potential method ambiguities.

**Pull request closed without merging**